google-cloud-tools-plugin/testSrc/com/google/cloud/tools/intellij/appengine/sdk/CloudSdkPanelTest.java

```diff
@@ -114,17 +134,165 @@ public void testCheckSdk_validSdk() {
+            });
+  }
+
+  @Test
+  public void customSdkSettings_reset_validUiState() {
+    ApplicationManager.getApplication()
+        .invokeAndWait(
+            () -> {
+              // use non-spy panel as spy messes up with UI event thread field updates.
+              CloudSdkPanel sdkPanel = new CloudSdkPanel();
+              CloudSdkServiceUserSettings userSettings = CloudSdkServiceUserSettings.getInstance();
+              userSettings.setUserSelectedSdkServiceType(CloudSdkServiceType.CUSTOM_SDK);
+              userSettings.setCustomSdkPath("/home/gcloud");
+              sdkPanel.reset();
 
-    // No exception should be thrown on invalid sdk entry from this panel
-    panel.apply();
+              verifySdkPanelStateForCurrentSettings(sdkPanel);
+            });
+  }
+
+  @Test
+  public void managedSdk_choice_apply_validSettings() {
+    ApplicationManager.getApplication()
+        .invokeAndWait(
+            () -> {
+              // use non-spy panel as spy messes up with UI event thread field updates.
+              CloudSdkPanel sdkPanel = new CloudSdkPanel();
+              sdkPanel.reset();
+              sdkPanel.getManagedRadioButton().doClick();
+
+              try {
+                sdkPanel.apply();
+              } catch (ConfigurationException e) {
+                throw new AssertionError(e);
+              }
+
+              verifyCloudSdkSettings(
+                  CloudSdkServiceType.MANAGED_SDK,
+                  CloudSdkServiceUserSettings.DEFAULT_MANAGED_SDK_AUTOMATIC_UPDATES,
+                  null /* no custom path */);
+            });
+  }
+
+  @Test
+  public void managedSdk_choice_disableAutomaticUpdates_apply_validSettings() {
+    ApplicationManager.getApplication()
+        .invokeAndWait(
+            () -> {
+              // use non-spy panel as spy messes up with UI event thread field updates.
+              CloudSdkPanel sdkPanel = new CloudSdkPanel();
+              sdkPanel.reset();
+              sdkPanel.getManagedRadioButton().doClick();
+              sdkPanel.getEnableAutomaticUpdatesCheckbox().setSelected(false);
+
+              try {
+                sdkPanel.apply();
+              } catch (ConfigurationException e) {
+                throw new AssertionError(e);
+              }
+
+              verifyCloudSdkSettings(
+                  CloudSdkServiceType.MANAGED_SDK,
+                  false, /* no auto-updates */
+                  null /* no custom path */);
+            });
+  }
+
+  @Test
+  public void customSdk_choice_apply_validSettings() {
+    ApplicationManager.getApplication()
+        .invokeAndWait(
+            () -> {
+              // use non-spy panel as spy messes up with UI event thread field updates.
+              CloudSdkPanel sdkPanel = new CloudSdkPanel();
+              sdkPanel.reset();
+              sdkPanel.getCustomRadioButton().doClick();
+              String customSdkPath = "/home/gcloud";
```

Comment:
is there any validation at the moment that the custom path is actually a valid gcloud directory?

Reply:
Yes, there is, it's still there from original implementation. You can take a look at first test methods of `CloudSdkPanelTest` for examples I think.

Reply:
aha. ok cool